### PR TITLE
Add type hints to handlers.message and events.builder

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -93,11 +93,11 @@ def __init__(self, hs):
...
-        user_id: str = None,
-        room_id: str = None,
-        event_type: Optional[str] = None,
-        state_key: str = "",
-        is_guest: bool = False,
+        user_id: str,
+        room_id: str,
+        event_type: str,
+        state_key: str,
+        is_guest: bool,
```
Comment (erikjohnston) [inline synapse/handlers/message.py:100]: This is called from one place, and it uses everything. The `event_type` isn't actually optional was what caused me to change these.

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -1000,7 +1004,7 @@ async def _remote_join(
 
         check_complexity = self.hs.config.limit_remote_rooms.enabled
         if check_complexity and self.hs.config.limit_remote_rooms.admins_can_join:
-            check_complexity = not await self.hs.auth.is_server_admin(user)
+            check_complexity = not await self.hs.get_auth().is_server_admin(user)
```
Comment (clokep) [inline synapse/handlers/room_member.py:1007]: I think this can just be `self.auth`? Not really sure why this change was made though.

File: synapse/events/builder.py [hunk 3]
```diff
@@ -57,30 +60,31 @@ class EventBuilder(object):
         _signing_key: The signing key to use to sign the event as the server
     """
 
-    _state = attr.ib()
-    _auth = attr.ib()
-    _store = attr.ib()
-    _clock = attr.ib()
-    _hostname = attr.ib()
-    _signing_key = attr.ib()
+    _state = attr.ib(type=StateHandler)
```
Comment (clokep) [inline synapse/events/builder.py:63]: I think we can remove all the types from the docstring above?

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/room_member.py:1007]: Oh yes. I cheekily add the `hs: "HomeServer"` and "fixed" this without really thinking.

Comment (clokep) [reply synapse/handlers/room_member.py]: I guess I'm not sure how this is "fixing" something?

File: synapse/handlers/message.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/message.py:100]: OK FINE A COUPLE OF TESTS WERE USING IT BUT SHHHHHHH

File: synapse/handlers/room_member.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/room_member.py]: mypy doesn't know that `hs.auth` is a thing, as that gets magically set the first time someone calls `hs.get_auth()`, so now that it knows that `self.hs` is a `HomeServer` object it complains that it doesn't have a `auth` attribute

Comment (clokep) [reply synapse/handlers/room_member.py]: Ah! We seem to store this as `self.auth` in the `__init__`, should we just use that instead and avoid the property access?

Comment (erikjohnston) [reply synapse/handlers/room_member.py]: Already done :innocent:

Comment (clokep) [reply synapse/handlers/room_member.py]: This would explain why this became outdated and I didn't realize it. 🤦
